tracer/test/Datadog.Trace.ClrProfiler.Managed.Tests/AutoInstrumentation/AWS/Kinesis/ContextPropagationTests.cs

```diff
@@ -0,0 +1,155 @@
+    private static readonly Dictionary<string, object> PokemonDictionary = new() { { "id", 393 }, { "name", "Piplup" }, { "type", "water" } };
+    private static readonly byte[] PersonJsonStringBytes = Encoding.UTF8.GetBytes(Vendors.Newtonsoft.Json.JsonConvert.SerializeObject(PersonDictionary));
+    private static readonly byte[] PersonBase64StringBytes = Convert.FromBase64String(Convert.ToBase64String(PersonJsonStringBytes));
+    private static readonly byte[] StreamNameBytes = Encoding.UTF8.GetBytes(StreamName);
+
+    private readonly SpanContext spanContext;
+
+    public ContextPropagationTests()
+    {
+        const long upper = 1234567890123456789;
+        const ulong lower = 9876543210987654321;
+
+        var traceId = new TraceId(upper, lower);
+        ulong spanId = 6766950223540265769;
+        spanContext = new SpanContext(traceId, spanId, 1, "test-kinesis", "serverless");
+    }
+
+    public static IEnumerable<object[]> MemoryStreamToDictionaryExpectedData
+        => new List<object[]>
+        {
+            new object[] { PersonBase64StringBytes, PersonDictionary },
+            new object[] { PersonJsonStringBytes, PersonDictionary },
+            new object[] { StreamNameBytes, null },
+        };
+
+    [Fact]
+    public void InjectTraceIntoRecords_WithJsonString_AddsTraceContext()
+    {
+        var request = new PutRecordsRequest { StreamName = StreamName, Records = new List<PutRecordsRequestEntry> { new PutRecordsRequestEntry { Data = ContextPropagation.DictionaryToMemoryStream(PersonDictionary), PartitionKey = Guid.NewGuid().ToString() }, new PutRecordsRequestEntry { Data = ContextPropagation.DictionaryToMemoryStream(PokemonDictionary), PartitionKey = Guid.NewGuid().ToString() } } };
+
+        var proxy = request.DuckCast<IPutRecordsRequest>();
+
+        ContextPropagation.InjectTraceIntoRecords<PutRecordsRequest>(proxy, spanContext);
+
+        var firstRecord = proxy.Records[0].DuckCast<IContainsData>();
+
+        // MemoryStreamToDictionary returns a Dictionary<string, object>
+        var dataDictionary = ContextPropagation.MemoryStreamToDictionary(firstRecord.Data);
+        var extracted = dataDictionary.TryGetValue(DatadogKey, out var datadogDictionary);
+        extracted.Should().BeTrue();
+
+        // Cast into a Dictionary<string, string> so we can extract it properly
+        var dictionary = JsonConvert.DeserializeObject<Dictionary<string, string>>(datadogDictionary.ToString());
+
+        var extractedSpanContext = SpanContextPropagator.Instance.Extract(dictionary);
+        extractedSpanContext.Should().NotBeNull();
+        extractedSpanContext.TraceId.Should().Be(spanContext.TraceId);
+        extractedSpanContext.SpanId.Should().Be(spanContext.SpanId);
+    }
+
+    [Fact]
+    public void InjectTraceIntoRecords_WithString_SkipsAddingTraceContext()
```

Comment:
I like the way this test method is implemented! I have two suggestions that I think could help make this test and the previous test better:
1. For this method and `InjectTraceIntoRecords_WithJsonString_AddsTraceContext`, it could be helpful to have a small helper method which takes in Data1 and Data2 to make it easier to show how the inputs differ for the two scenarios.
2. For `InjectTraceIntoRecords_WithJsonString_AddsTraceContext`, I think it would be a stronger test if it didn't invoke any of the extraction logic in ContextPropagation, since then the test begins to test both inject and extract, instead of just inject. I think this could be done by extracting the entire data string from the MemoryStream (which would be a JSON string in that case), using the Newtonsoft JSON deserializer to convert it into JObjects, and then asserting that the dictionary entries `jObject["datadog"]`, `jObject["datadog"]["x-datadog-trace-id"]`, and `jObject["datadog"]["x-datadog-parent-id"]` are what we expect.

I wonder what other teammates think, it's possible I'm overthinking the second suggestion.

Reply:
Followed your suggestion after our chat around this.